tests/many_to_many/tests.py

```diff
@@ -554,3 +560,27 @@ def test_inherited_models_selects(self):
             ]
         )
         self.assertQuerysetEqual(b.publications.all(), ['<Publication: Science Weekly>'])
+
+    def test_count_and_exists_query(self):
+        """
+        #29725 - Calling count() and exists() on a many to many relation
+        should not generate sql including unnecessary JOIN.
+        """
+        article = self.a1
+        nullable_target_article = NullableTargetArticle.objects.create(headline='Python is good')
+        NullablePublicationThrough.objects.create(article=nullable_target_article, publication=None)
+        NullablePublicationThrough.objects.create(article=nullable_target_article, publication=self.p1)
+
+        with CaptureQueriesContext(connection) as captured_query:
+            self.assertEqual(article.publications.count(), 1)
+            self.assertNotIn('JOIN', captured_query[0]['sql'])
+
+            self.assertEqual(nullable_target_article.publications.count(), 1)
+
+            self.assertTrue(article.publications.exists())
+            self.assertNotIn('JOIN', captured_query[2]['sql'])
+
+            self.assertTrue(nullable_target_article.publications.exists())
+
+        with mock.patch.object(connection.features, 'supports_foreign_keys', False):
+            self.assertIsNone(article.publications.constrained_target)
```

Comment:
I would personally have broken this test in two (`test_count_join_optimization`, `test_exist_join_optimization`) and asserted against the results instead of testing the `constrained_target` implementation details but I guess this works fine.

Reply:
Thanks to your reviews @charettes : )
But, I don't understand how to assert against the results.
Is it OK just to check that 'JOIN' is included in the `captured_query` when `supports_foreign_keys` is `False` in this context?
Or is there any better way?

Reply:
Oh I realize that asserting against the results is problematic given all the engines we're testing against support foreign keys. In this case yes, using the same `JOIN` check against `captured_query` should do!

Reply:
Thanks @charettes 
I fixed the tests according to your feedbacks

Reply:
Thanks @charettes 
I fixed the tests according to your advice.